File: leetcode/70.rs
/*
You are climbing a staircase. It takes n steps to reach the top.
Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
*/

// Approach:
// The answer is just the nth value in the fibonacci series 
// So we just run a loop to calculate fibonacci series upto n and return the value

impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut a = 0;
        let mut b = 1;
        for i in 0..n {
            let tmp = b;
            b = a+b;
            a = tmp;
        }
        b
    }
}

File: leetcode/70.java
/*
Question:
You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

*/

// Approach:
// 1) If n is 1, return 1
// 2) Create a dp array of size n
// 3) Initialize dp[0] = 1, dp[1] = 2
// 4) For the rest of the array, dp[i] = dp[i-1] + dp[i-2]
// 5) Return dp[n-1]

class Solution {
  public int climbStairs(int n) {
    if(n == 1) return 1;
    int[] dp = new int[n];
    dp[0] = 1;
    dp[1] = 2;
    for(int i = 2; i < n; i++) {
      dp[i] = dp[i-1] + dp[i-2];
    }
    return dp[n-1];
  }
}
